ably/realtime_presence.go

```diff
@@ -100,23 +130,60 @@ func (pres *RealtimePresence) syncWait() {
 	pres.syncMtx.Unlock()
 }
 
-func syncSerial(msg *protocolMessage) string {
-	if i := strings.IndexRune(msg.ChannelSerial, ':'); i != -1 {
-		return msg.ChannelSerial[i+1:]
+// RTP18
+func syncSerial(msg *protocolMessage) (noChannelSerial bool, syncSequenceId string, syncCursor string) {
+	if empty(msg.ChannelSerial) { // RTP18c
+		noChannelSerial = true
+		return
+	}
+	// RTP18a
+	serials := strings.Split(msg.ChannelSerial, ":")
+	syncSequenceId = serials[0]
+	if len(serials) > 1 {
+		syncCursor = serials[1]
 	}
-	return ""
+	return false, syncSequenceId, syncCursor
 }
 
-func (pres *RealtimePresence) onAttach(msg *protocolMessage) {
-	serial := syncSerial(msg)
+func (pres *RealtimePresence) enterMembers(internalMembers []*PresenceMessage) {
+	for _, member := range internalMembers {
+		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Minute)
```

Comment:
Is there any context as to why this timeout is set to 2 minutes?

Reply:
Honestly there's no context as such, I just didn't want goroutine to get blocked!
Your suggestions will be helpful here !!

Reply:
Perhaps it should keep retrying whilst the channel is in a state where we can expect the operation to eventually succeed (i.e. it's not in a terminal state)?

Reply:
I just don't want goroutine to go into a dangling state : (
Sure, it can work.

Reply:
> To put another way, what if the enterClient call is failing due to an unreliable network for 3 minutes, but then it's ok again, shouldn't this then succeed rather than have give up after 2 minutes?

If a message is sent to the connection while it's in a `connected` state, then eventually either it will successfully send or the connection will detect a network issue and become disconnected. If the latter happens, then when it reconnects, the channel will re-attach, which will trigger `onAttach()` and `presence.enterMembers()` to happen again at that time. So we shouldn't need an old enterMembers from some previous attach cycle living on.

Reply:
Sorry yes, I'm proposing `context.Background()`, since if I understand Simon correctly, the call will either block until it succeeds, or return an error if the client becomes disconnected, so in effect there is a timeout on this operation based on how long it takes the client to detect a disconnection, there's no need to specify an arbitrary timeout here too.

Reply:
Same will happen with `cancelledContext` too. If it fails to send message due to `disconnection`, for next channel attach it will automatically send the message. We don't really care about message failure in current case. Because, it will only fail when channel is in some other state ( maybe because of disconnection ), for next attach, we can enter members again.
There's no need to wait really

Reply:
Update -> `context.Background()` does seem to be a sensible approach, but is there a way this can go into a lock type of a situation where client is connected for a long time and ack is not received for the entered presence.

Reply:
That's why I tried putting a `timeout` here, where all `innerPresenceMembers` will be processed without one waiting for another

Reply:
Ok, I leave it to your judgement, I'd probably just use `context.Background()`